src/gfx/pal_packing.cpp

```diff
@@ -0,0 +1,359 @@
+
+	struct Iota {
+		using value_type = size_t;
+		using difference_type = size_t;
+		using pointer = value_type const *;
+		using reference = value_type const &;
+		using iterator_category = std::input_iterator_tag;
+
+		// Use aggregate init etc.
+		value_type i;
+
+		bool operator!=(Iota const &other) { return i != other.i; }
+		reference operator*() const { return i; }
+		pointer operator->() const { return &i; }
+		Iota operator++() {
+			++i;
+			return *this;
+		}
+		Iota operator++(int) {
+			Iota copy = *this;
+			++i;
+			return copy;
+		}
+	};
+
+	// Begin with all proto-palettes queued up for insertion
+	std::queue queue(std::deque<ProtoPalAttrs>(Iota{0}, Iota{protoPalettes.size()}));
+	// Begin with no pages
+	std::vector<AssignedProtos> assignments{};
+
+	while (!queue.empty()) {
+		ProtoPalAttrs const &attrs = queue.front();
+		ProtoPalette const &protoPal = protoPalettes[attrs.palIndex];
+		size_t bestPalIndex = assignments.size();
+		// We're looking for a palette where the proto-palette's relative size is less than
+		// its actual size; so only overwrite the "not found" index on meeting that criterion
+		double bestRelSize = protoPal.size();
+
+		for (size_t i = 0; i < assignments.size(); ++i) {
+			// Skip the page if this one is banned from it
+			if (attrs.isBannedFrom(i)) {
+				continue;
+			}
+
+			options.verbosePrint("%zu: Rel size: %f (size = %zu)\n", i,
+			                     assignments[i].relSizeOf(protoPal), protoPal.size());
+			if (assignments[i].relSizeOf(protoPal) < bestRelSize) {
+				bestPalIndex = i;
+			}
+		}
+
+		if (bestPalIndex == assignments.size()) {
+			// Found nowhere to put it, create a new page containing just that one
+			assignments.emplace_back(protoPalettes, std::move(attrs));
+		} else {
+			auto &bestPal = assignments[bestPalIndex];
+			// Add the color to that palette
+			bestPal.assign(std::move(attrs));
+
+			// If this overloads the palette, get it back to normal (if possible)
+			while (bestPal.volume() > 4) {
+				options.verbosePrint("Palette %zu is overloaded! (%zu > 4)\n", bestPalIndex,
+				                     bestPal.volume());
+
+				// Look for a proto-pal minimizing "efficiency" (size / rel_size)
+				auto efficiency = [&bestPal](ProtoPalette const &pal) {
+					return pal.size() / bestPal.relSizeOf(pal);
+				};
+				auto [minEfficiencyIter, maxEfficiencyIter] =
+					std::minmax_element(bestPal.begin(), bestPal.end(),
+				                        [&efficiency, &protoPalettes](ProtoPalAttrs const &lhs,
+				                                                      ProtoPalAttrs const &rhs) {
+											return efficiency(protoPalettes[lhs.palIndex])
+					                               < efficiency(protoPalettes[rhs.palIndex]);
+										});
+
+				// All efficiencies are identical iff min equals max
+				// TODO: maybe not ideal to re-compute these two?
+				// TODO: yikes for float comparison! I *think* this threshold is OK?
```

Comment:
Let's say `protoPalettes[maxEfficiencyIter->palIndex]` is `A` and `protoPalettes[minEfficiencyIter->palIndex]` is `B`. This compares `efficiency(A)` and `efficiency(B)` with a threshold. But `efficiency(x) = x.size() / bestPal.relSizeOf(x)`; so instead, how about finding the LCM of `bestPal.relSizeOf(A)` and `bestPal.relSizeOf(B)` and comparing `A.size()` and `B.size()` based on that?

Reply:
Can you explain that a bit more?

Reply:
```
A := protoPalettes[maxEfficiencyIter->protoPalIndex]
B := protoPalettes[minEfficiencyIter->protoPalIndex]
condition := efficiency(A) ==~ efficiency(B)
condition := A.size() / bestPal.relSizeOf(A) ==~ B.size() / bestPal.relSizeOf(B) // float threshold
condition := A.size() * bestPal.relSizeOf(B) == B.size() * bestPal.relSizeOf(A) // exact comparison
```

Reply:
Ah, I meant regarding how the two algorithms are equivalent, sorry

Reply:
I just multiplied both sides by the denominators.

Reply:
```c++
auto efficiency = [&bestPal](ProtoPalette const &pal) {
	return pal.size() / bestPal.relSizeOf(pal);
};
```

Reply:
Oh, right. Forgive me for lacking in confidence, but I'd like to implement this change after we've set up a test corpus, so we can ensure that this does not accidentally changes generation output.

Reply:
Ah, this doesn't work because `relSizeOf` is floating-point as well. Any other ideas?